system/BaseModel.php

```diff
@@ -0,0 +1,1832 @@
+	 */
+	protected abstract function dbFind(bool $singleton, $id = null);
+
+	/**
+	 * Fetches the column of database from $this->table
+	 *
+	 * @param string $columnName Column Name
+	 *
+	 * @return array|null   The resulting row of data, or null if no data found.
+	 * @throws DataException Data Exception.
+	 */
+	public function findColumn(string $columnName)
+	{
+		if (strpos($columnName, ',') !== false)
+		{
+			throw DataException::forFindColumnHaveMultipleColumns();
+		}
+
+		$resultSet = $resultSet = $this->dbFindColumn($columnName);
+
+		return $resultSet ? array_column($resultSet, $columnName) : null;
+	}
+
+	/**
+	 * Fetches the column of database from $this->table
+	 *
+	 * @param string $columnName Column Name
+	 *
+	 * @return array|null   The resulting row of data, or null if no data found.
+	 * @throws DataException Data Exception.
+	 */
+	protected abstract function dbFindColumn(string $columnName);
+
+	/**
+	 * Works with the current Query Builder instance to return
+	 * all results, while optionally limiting them.
+	 *
+	 * @param integer $limit  Limit
+	 * @param integer $offset Offset
+	 *
+	 * @return array
+	 */
+	public function findAll(int $limit = 0, int $offset = 0)
+	{
+		if ($this->tempAllowCallbacks)
+		{
+			// Call the before event and check for a return
+			$eventData = $this->trigger('beforeFind', [
+				'method'    => 'findAll',
+				'limit'     => $limit,
+				'offset'    => $offset,
+				'singleton' => false,
+			]);
+
+			if (! empty($eventData['returnData']))
+			{
+				return $eventData['data'];
+			}
+		}
+
+		$eventData = [
+			'data'      => $this->dbFindAll($limit, $offset),
+			'limit'     => $limit,
+			'offset'    => $offset,
+			'method'    => 'findAll',
+			'singleton' => false,
+		];
+
+		if ($this->tempAllowCallbacks)
+		{
+			$eventData = $this->trigger('afterFind', $eventData);
+		}
+
+		$this->tempReturnType     = $this->returnType;
+		$this->tempUseSoftDeletes = $this->useSoftDeletes;
+		$this->tempAllowCallbacks = $this->allowCallbacks;
+
+		return $eventData['data'];
+	}
+
+	/**
+	 * Works with the current Query Builder instance to return
+	 * all results, while optionally limiting them.
+	 *
+	 * @param integer $limit  Limit
+	 * @param integer $offset Offset
+	 *
+	 * @return array
+	 */
+	protected abstract function dbFindAll(int $limit = 0, int $offset = 0);
+
+	/**
+	 * Returns the first row of the result set. Will take any previous
+	 * Query Builder calls into account when determining the result set.
+	 *
+	 * @return array|object|null
+	 */
+	public function first()
+	{
+		if ($this->tempAllowCallbacks)
+		{
+			// Call the before event and check for a return
+			$eventData = $this->trigger('beforeFind', [
+				'method'    => 'first',
+				'singleton' => true,
+			]);
+
+			if (! empty($eventData['returnData']))
+			{
+				return $eventData['data'];
+			}
+		}
+
+		$eventData = [
+			'data'      => $this->dbFirst();,
```

Comment:
```suggestion
			'data'      => $this->dbFirst(),
```
My fault copy paste failure